### PR TITLE
Add Column Roles for Google Charts DataTable

### PR DESCRIPTION
**Pull request by armandocanals**
This adds the ability to define a column role, or multiple roles (tooltip text, data point annotations, uncertainty indicators) for series data. I'm not sure if this is the right format for this feature, but calling it looks like this:

```
var data = [..., [Sat Mar 05 2016 17:13:22 GMT-0800 (PST), 1, false, "cool annotation"]];
new Chartkick.LineChart("chart-1", data, {"library": {"role":["certainty", "annotation"]}});
```

Outputs:

screen shot 2016-03-05 at 5 21 32 pm

**Potential Caveats** 

1) Expects the series data to include the data role columns in the same order as described in the library options:

```
[Sat Mar 05 2016 17:13:22 GMT-0800 (PST), 1, false, "cool annotation"]
```

What this means is that since the value for certainty (boolean) is specified before the value for annotation (string) in the library options, the series data is expected to contain the role values in that order.

2) Not all column roles are supported by all chart types. More info about Column Roles: https://developers.google.com/chart/interactive/docs/roles

_I also want to say thanks for all the hard work on this library (and others) :)_

### CODE REVIEW

File: chartkick.js [hunk 1]
```diff
@@ -793,10 +835,19 @@
...
-    var r = [], key, j;
-    for (j = 0; j < data.length; j++) {
-      key = toFormattedKey(data[j][0], keyType);
-      r.push([key, toFloat(data[j][1])]);
+    var r = [], key, i, j, datum;
+    for (i = 0; i < data.length; i++) {
+      key = toFormattedKey(data[i][0], keyType);
+      datum = [key];
+      for(j = 0; j < data[i].length; j++) {
+        if (j === 0) continue;
```
Comment (ggilder) [inline chartkick.js]: just start the `for` loop at 1?

File: chartkick.js [hunk 2]
```diff
@@ -555,9 +555,26 @@
 
         var jsOptions = jsOptionsFunc(defaultOptions, hideLegend, setMin, setMax, setStacked, setXtitle, setYtitle);
 
+        var getRoleDataType = function (name) {
+          var r = {
```
Comment (ggilder) [inline chartkick.js]: This seems like it should be moved out of function scope; no need to create the object every time the function is called

File: chartkick.js [hunk 3]
```diff
@@ -555,9 +555,26 @@
...
+
+          return r[name] || false;
+        };
+
         // cant use object as key
-        var createDataTable = function (series, columnType) {
-          var i, j, s, d, key, rows = [];
+        var createDataTable = function (series, columnType, role) {
+          var i, j, s, d, r, itm, key, rows = [];
+          var role = role || false;
```
Comment (ggilder) [inline chartkick.js]: Don't think it's necessary to coerce falsey values to `false` — this var is only used in the conditional below so this makes no difference

File: chartkick.js [hunk 4]
```diff
@@ -567,7 +584,18 @@
               if (!rows[key]) {
                 rows[key] = new Array(series.length);
               }
-              rows[key][i] = toFloat(d[1]);
+              r = [toFloat(d[1])];
+
+              if (role) {
+                for(itm = 0; itm < d.length; itm++) {
+                  if (itm != 0 && itm != 1) {
```
Comment (ggilder) [inline chartkick.js]: just start the for loop at 2? and maybe comment why...

**Comment by armandocanals** [comment]
@ggilder Just updated the PR addressing your notes. Thanks for the helpful feedback!

**Comment by ChrisCPO** [comment]
What's the status of merging this feature?

**Comment by stvmachine** [comment]
+1

**Comment by erickveras** [comment]
+1

**Comment by ankane** [comment]
Hey @armandocanals, thanks for the PR 👍 and really sorry for not responding earlier. It's a neat feature, but not something I want to maintain right now.

**Pull request closed without merging**